File: problems/102.binary-tree-level-order-traversal.py
#
# @lc app=leetcode id=102 lang=python3
#
# [102] Binary Tree Level Order Traversal
#

from typing import Optional, List, Tuple
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        output: List[List[int]] = []

        if not root:
            return output
        
        q = deque[Tuple[int, TreeNode]]()
        q.append((0, root))
        while (len(q) > 0):
            layer, current = q.pop()

            if layer >= len(output):
                output.append([])

            output[layer].append(current.val)

            if current.right:
                q.append((layer + 1, current.right))
            if current.left:
                q.append((layer + 1, current.left))

        return output

# @lc code=end

File: problems/102.binary-tree-level-order-traversal.cpp
/*
 * @lc app=leetcode id=102 lang=cpp
 *
 * [102] Binary Tree Level Order Traversal
 */

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> output;
        if (root == nullptr) {
            return output;
        }

        queue<pair<int, TreeNode*>> q;
        q.push({0, root});
        while (!q.empty()) {
            auto [layer, current] = q.front();
            q.pop();

            if (layer >= output.size()) {
                output.push_back({});
            }

            output[layer].push_back(current->val);

            if (current->left != nullptr) q.push({layer + 1, current->left});
            if (current->right != nullptr) q.push({layer + 1, current->right});
        }

        return output;
    }
};
// @lc code=end

